File: location/uikit/map.go
// Package uikit provides presentation-only models for 420Location map consumers.
// It never queries a provider or upgrades an approximate/private location to an exact pin.
package uikit

import (
 "errors"
 "math"
 "sort"
 "strings"

 "github.com/420integrated/420-integrated/location/model"
)

const MaxMapItems = 500

type Kind string
const (KindPin Kind = "pin"; KindArea Kind = "area")

type Item struct {
 ID string `json:"id"`
 Name string `json:"name"`
 Category model.Category `json:"category"`
 Kind Kind `json:"kind"`
 Latitude *float64 `json:"latitude,omitempty"`
 Longitude *float64 `json:"longitude,omitempty"`
 Region string `json:"region,omitempty"`
 City string `json:"city,omitempty"`
 Country string `json:"country,omitempty"`
}

type View struct { Items []Item `json:"items"`; Empty bool `json:"empty"` }

// Build accepts canonical places, emits public locations only and never emits
// coordinates for approximate locations (even if the source has coordinates).
// An invalid public record fails closed instead of leaking partially validated data.
func Build(places []model.Place)(View,error) {
 if len(places)>MaxMapItems {return View{},errors.New("map item limit exceeded")}
 out:=make([]Item,0,len(places))
 seen:=make(map[string]struct{},len(places))
 for _,p:=range places {
  if p.Visibility!=model.VisibilityPublic {continue}
  if err:=p.Validate();err!=nil{return View{},err}
  if _,ok:=seen[p.ID];ok{return View{},errors.New("duplicate public place id")};seen[p.ID]=struct{}{}
  item:=Item{ID:p.ID,Name:p.Name,Category:p.Category,Kind:KindArea,City:p.City,Region:p.Region,Country:p.Country}
  if p.Precision==model.PrecisionExactPublic {
   if p.Latitude==nil||p.Longitude==nil||math.IsNaN(*p.Latitude)||math.IsNaN(*p.Longitude) {return View{},errors.New("public pin coordinates missing")}
   lat,lon:=*p.Latitude,*p.Longitude
   item.Kind=KindPin;item.Latitude=&lat;item.Longitude=&lon
  } else {
   // Approximate entries expose only coarse textual geography, no street address,
   // original or inferred position, postal region, or provider alias.
   if strings.TrimSpace(item.City)==""&&strings.TrimSpace(item.Region)==""&&strings.TrimSpace(item.Country)=="" {continue}
  }
  out=append(out,item)
 }
 sort.Slice(out,func(i,j int)bool{if out[i].Name!=out[j].Name{return out[i].Name<out[j].Name};return out[i].ID<out[j].ID})
 return View{Items:out,Empty:len(out)==0},nil
}

File: location/uikit/map.js
// GEN-SVC-2.17: dependency-free, presentation-only 420Location map/list kit.
// `view` must be produced by the trusted location/uikit.Build projection on the server.
// An injected renderer can be MapLibre or any compatible provider. No geolocation,
// tile URLs, access tokens, wallets, or provider requests are embedded in this kit.
export function mountLocationMap(root, view, { renderer = null, onSelect = () => {} } = {}) {
  if (!(root instanceof HTMLElement)) throw new TypeError('a root element is required');
  if (!view || !Array.isArray(view.items) || view.items.length > 500) throw new TypeError('invalid map view');
  if (typeof onSelect !== 'function') throw new TypeError('onSelect must be a function');
  const items = view.items.map(item => {
    if (!item || typeof item.id !== 'string' || !item.id || typeof item.name !== 'string' || !item.name) throw new TypeError('invalid map item');
    if (item.kind !== 'pin' && item.kind !== 'area') throw new TypeError('invalid map precision');
    if (item.kind === 'area' && (item.latitude != null || item.longitude != null)) throw new TypeError('approximate location includes coordinates');
    if (item.kind === 'pin' && (!Number.isFinite(item.latitude) || !Number.isFinite(item.longitude) || Math.abs(item.latitude) > 90 || Math.abs(item.longitude) > 180)) throw new TypeError('invalid exact pin');
    return { ...item };
  });
  root.replaceChildren();
  const container = document.createElement('section');
  container.className = 'location-map-kit';
  container.setAttribute('aria-label', 'Places map and list');
  const heading = document.createElement('h2'); heading.textContent = 'Explore places'; container.append(heading);
  const status = document.createElement('p'); status.setAttribute('role', 'status'); status.setAttribute('aria-live', 'polite');
  const list = document.createElement('ul'); list.className = 'location-map-list'; list.setAttribute('aria-label', 'Places');
  const pins = items.filter(item => item.kind === 'pin');
  const mapRegion = document.createElement('div'); mapRegion.className = 'location-map-canvas'; mapRegion.setAttribute('aria-label', 'Public place map');
  const buttons = new Map();
  let selected = null;
  for (const item of items) {
    const li = document.createElement('li');
    const button = document.createElement('button'); button.type = 'button'; button.textContent = item.name;
    button.setAttribute('aria-pressed', 'false');
    const subtitle = document.createElement('span'); subtitle.className = 'location-map-detail';
    subtitle.textContent = [item.category, item.city, item.region, item.country].filter(Boolean).join(' · ');
    if (item.kind === 'area') subtitle.textContent += ' · approximate area';
    button.addEventListener('click', () => select(item.id));
    li.append(button, subtitle); list.append(li); buttons.set(item.id, button);
  }
  function select(id) {
    const item = items.find(candidate => candidate.id === id);
    if (!item) return;
    selected = id;
    for (const [key, button] of buttons) button.setAttribute('aria-pressed', key === id ? 'true' : 'false');
    status.textContent = `Selected ${item.name}`;
    onSelect({ ...item });
  }
  let disposeRenderer = () => {};
  if (renderer && typeof renderer.mount === 'function' && pins.length) {
    // Never pass approximate-area coordinates to a renderer.
    const result = renderer.mount(mapRegion, pins.map(item => ({ ...item })), { select });
    if (typeof result === 'function') disposeRenderer = result;
    container.append(mapRegion);
  } else {
    status.textContent = 'List view available; map renderer not configured.';
  }
  if (!items.length) status.textContent = 'No public places to display.';
  container.append(status, list); root.append(container);
  return { select, getSelected: () => selected, destroy: () => { disposeRenderer(); root.replaceChildren(); } };
}
